### ISSUE TITLE
implement the ack-frequency extensions

### ISSUE THREAD

**Issue by marten-seemann** [lucas-clemente/quic-go#2482]
Wire format changes of this extensions are now tied to the QUIC drafts, so we can safely implement and deploy this extension.

https://janaiyengar.github.io/ack-frequency/draft-iyengar-quic-delayed-ack.html

**Comment by vikulin** [comment]
Any updates on this?

**Comment by vikulin** [comment]
Adding https://datatracker.ietf.org/doc/html/draft-iyengar-quic-delayed-ack-00

**Comment by marten-seemann** [comment]
That's not the correct link: https://datatracker.ietf.org/doc/html/draft-ietf-quic-ack-frequency

We won't implement this before until this document has made significant progress at the IETF (possibly WGLC?). Even then, we'd need a test suite able to demonstrate the benefit of this extension.

**Comment by vikulin** [comment]
> That's not the correct link: https://datatracker.ietf.org/doc/html/draft-ietf-quic-ack-frequency
> 
> We won't implement this before until this document has made significant progress at the IETF (possibly WGLC?). Even then, we'd need a test suite able to demonstrate the benefit of this extension.

yes, you are right. Can you please point out what ack frequency is currently applied for quic-go? I found this line but I am not sure it is correct:
https://github.com/lucas-clemente/quic-go/blob/master/internal/ackhandler/sent_packet_handler.go#L732

**Comment by marten-seemann** [comment]
quic-go implements the acknowledgement strategy described in RFC 9002.

**Comment by vikulin** [comment]
> > That's not the correct link: https://datatracker.ietf.org/doc/html/draft-ietf-quic-ack-frequency
> > We won't implement this before until this document has made significant progress at the IETF (possibly WGLC?). Even then, we'd need a test suite able to demonstrate the benefit of this extension.
> 
> yes, you are right. Can you please point out what ack frequency is currently applied for quic-go? I found this line but I am not sure it is correct: https://github.com/lucas-clemente/quic-go/blob/master/internal/ackhandler/sent_packet_handler.go#L732

I found following: 
```
A receiver SHOULD send an ACK frame after receiving at least two ack-eliciting packets
```
@marten-seemann could you please point out where the rule is implemented in quic-go code? This is only what I am asking for.

**Comment by marten-seemann** [comment]
https://github.com/lucas-clemente/quic-go/blob/master/internal/ackhandler/received_packet_tracker.go

**Comment by larryliu2018** [comment]
Hi Marten Seemann
I tested quic-go on the 100Mbps 10ms link simulated by mininet, and set const packetsBeforeAck = 2 at the same time, but the ratio of the number of ack packets calculated by qlog to the number of received packets is not 1:2, , and the difference is huge.

Below is the result of my test setup with different bandwidths and with packetsBeforeAck = 2, 4, 8, 16. The transfer file size is 1MB, I don't know why, quic-go does not send ack as expected, I am using the client and server programs under interop in quic-go-v0.27.0, I want to ask you what Caused by? I really hope for your help, thank you
![ack-bw-change](https://user-images.githubusercontent.com/35099991/170405354-c374dc3b-e02e-409e-b92f-ee3c9698503f.jpg)
![data-bw-change](https://user-images.githubusercontent.com/35099991/170405358-a23411f8-a9c2-4b23-954f-b40ec5dd3b83.jpg)

**Comment by marten-seemann** [comment]
You're seeing ACKs sent less than for every 2 packets. That's expected. For example, if we receive 5 packets at basically the same time (i.e. before we manage to pack an acknowledgement packet), we'd just acknowledge them in one packet.

**Comment by larryliu2018** [comment]
Thank you very much for your answer. You mean that packetsBeforeAck defines a lower bound. If the bandwidth of the link is too large, multiple data packets will arrive almost at the same time before ack packaging, and multiple data packets will be confirmed at once, but this number will be Greater than or equal to packetsBeforeAck. Then the code you designed has a certain degree of adaptability to the acknowledgment frequency of the received data packets. I originally wanted to use ACK_FREQUENCY frames to do some optimization work. Especially in high-bandwidth scenarios, the data link throughput may decrease due to the high sending frequency of acks, which is not necessary now.
The figure below is the ack sending scenario I simulated on wifi, but using quic-go does not show similar problems, or the throughput drop is not obvious

![data](https://user-images.githubusercontent.com/35099991/170465706-1aca596c-fd75-4de4-9e0d-151f96cda999.jpg)

**Comment by marten-seemann** [comment]
> I originally wanted to use ACK_FREQUENCY frames to do some optimization work. Especially in high-bandwidth scenarios, the data link throughput may decrease due to the high sending frequency of acks, which is not necessary now.

Yes, this is exactly what this extension is for. It definitely needs to be adaptive, sending one ACK every 16 packets is probably too little for slower conditions.
Any idea what a good heuristic would look like?

**Comment by vikulin** [comment]
> > I originally wanted to use ACK_FREQUENCY frames to do some optimization work. Especially in high-bandwidth scenarios, the data link throughput may decrease due to the high sending frequency of acks, which is not necessary now.
> 
> Yes, this is exactly what this extension is for. It definitely needs to be adaptive, sending one ACK every 16 packets is probably too little for slower conditions. Any idea what a good heuristic would look like?

This [article ](https://www.fastly.com/blog/measuring-quic-vs-tcp-computational-efficiency) shows some input data.

**Comment by larryliu2018** [comment]
> > I originally wanted to use ACK_FREQUENCY frames to do some optimization work. Especially in high-bandwidth scenarios, the data link throughput may decrease due to the high sending frequency of acks, which is not necessary now.
> 
> Yes, this is exactly what this extension is for. It definitely needs to be adaptive, sending one ACK every 16 packets is probably too little for slower conditions. Any idea what a good heuristic would look like?

There is a SIGCOMM'20 article here, it balances byte-counting ACK and periodic ACK so as to achieve a controlled ACK frequency, which feels valuable for reference, but he has made a lot of changes to the protocol.
TACK: Improving Wireless Transport Performance by Taming Acknowledgments
https://dl.acm.org/doi/abs/10.1145/3387514.3405850

**Comment by larryliu2018** [comment]
Hello, I recently looked at how to modify the sending frequency of acks. I have a question that I would like to ask you.
Function defined in file "quic-go/internal/ackhandler/received_packet_tracker.go": func (h *receivedPacketTracker) maybeQueueAck(pn protocol.PacketNumber, rcvTime time.Time, wasMissing bool) ,

Among them, there is a variable h.ackQueued. If this parameter is false, as long as the timer of delay ack does not time out, will it not send ACK? Can I control the sending frequency of ACK by controlling this parameter?

In addition, I also found that on a link with a packet loss rate of 10%, the throughput of the quic-go test was reduced by 90%, and the congestion control algorithm used was cubic. I don't know if there is a problem with my experiment. Have you ever had a similar test experience? @marten-seemann

**Comment by larryliu2018** [comment]
Hello, I recently looked at how to modify the sending frequency of acks. I have a question that I would like to ask you.
Function defined in file "quic-go/internal/ackhandler/received_packet_tracker.go": func (h *receivedPacketTracker) maybeQueueAck(pn protocol.PacketNumber, rcvTime time.Time, wasMissing bool) ,

Among them, there is a variable h.ackQueued. If this parameter is false, as long as the timer of delay ack does not time out, will it not send ACK? Can I control the sending frequency of ACK by controlling this parameter?

In addition, I also found that on a link with a packet loss rate of 10%, the throughput of the quic-go test was reduced by 90%, and the congestion control algorithm used was cubic. I don't know if there is a problem with my experiment. Have you ever had a similar test experience? @marten-seemann

**Comment by larryliu2018** [comment]
Hi Marten Seemann
I tested quic-go on the 100Mbps 10ms link simulated by mininet, no packet loss is set. The client and server programs under the intorop file used, transmit 100MB files, and set const packetsBeforeAck = 2 at the same time. However, the test results show that there are packet losses.

![Image 19](https://user-images.githubusercontent.com/35099991/172603414-851b290f-b2cc-4817-b448-22f171801b67.png)
In Figure 1, the line-by-line parameters output by the client are, respectively, goodput, the number of sent packets, the number of received packets, the number of declared lost packets, the number of ACK packets sent, and the number of ACK packets received. There is no goodput line on the server side, these parameters are counted by the program in qlog. Among them, goodput needs to be multiplied by 100, because the code calculation is calculated by dividing the 1MB file sent by the statistical transmission completion time, and the actual sending is 100MB.


![Image 20](https://user-images.githubusercontent.com/35099991/172604017-945f8344-7d26-4280-af73-526dce4617e1.png)
![Image 21](https://user-images.githubusercontent.com/35099991/172604048-f49b3908-2a04-4493-920d-581132edacb6.png)
In Figure 2-Figure 3, use qvis to view the generated qlog file. Checking also shows packet loss.
I want to ask you why? is this normal?


After I set a larger value of packetsBeforeAck and reduce the frequency of ACK, serious packet loss also occurred. I really hope for your help, thank you.

@marten-seemann

**Comment by vikulin** [comment]
Got it, Thanks

пн, 23 мая 2022 г. в 18:14, Marten Seemann ***@***.***>:

**Comment by marten-seemann** [comment]
draft-03 has shipped: https://datatracker.ietf.org/doc/draft-ietf-quic-ack-frequency/03/
And it looks like [all issues](https://github.com/quicwg/ack-frequency) have been resolved. Maybe this draft is soon ready for WGLC? That would be good time to update my branch and get this merged. We'll probably get some more news in [Yokohama](https://www.ietf.org/how/meetings/116/).

**Comment by marten-seemann** [comment]
See Ian's presentation at IETF 117 for some takeaways from early mvfst deployment: https://datatracker.ietf.org/meeting/117/materials/slides-117-quic-ack-frequency

**Comment by marten-seemann** [comment]
The draft is now in WGLC. This would be a good time to implement it, however, I'm not sure if I can prioritize it at this point.
